src/storage/index/hnsw_index.cpp

```diff
@@ -339,37 +338,272 @@ common::offset_t OnDiskHNSWIndex::searchNNInUpperLayer(transaction::Transaction*
+    } break;
+    default: {
+        KU_UNREACHABLE;
+    }
+    }
+    return neighbors;
+}
+
+common::offset_vec_t OnDiskHNSWIndex::oneHopFilteredSearch(common::SemiMask* semiMask, uint64_t ef,
+    graph::Graph::EdgeIterator& nbrItr, VisitedState& visited) {
+    common::offset_vec_t result;
+    result.reserve(ef);
+    for (const auto& neighborChunk : nbrItr) {
+        neighborChunk.forEach([&](auto neighbor, auto) {
+            if (!visited.contains(neighbor.offset)) {
+                if (semiMask->isMasked(neighbor.offset)) {
+                    visited.add(neighbor.offset);
+                    result.push_back(neighbor.offset);
+                }
+            }
+        });
+    }
+    return result;
+}
+
+common::offset_vec_t OnDiskHNSWIndex::directedTwoHopFilteredSearch(
+    transaction::Transaction* transaction, common::SemiMask* semiMask, const float* queryVector,
+    uint64_t ef, graph::Graph::EdgeIterator& nbrItr, VisitedState& visited,
+    NodeTableScanState& embeddingScanState, graph::NbrScanState& secondHopScanState,
+    min_node_priority_queue_t& candidates, max_node_priority_queue_t& results) const {
+    min_node_priority_queue_t candidatesForSecHop;
+
+    uint64_t numNbrsInFirstHop = 0;
+    uint64_t numVisitedNbrs = 0;
+    for (const auto& neighborChunk : nbrItr) {
+        neighborChunk.forEach([&](auto neighbor, auto) {
+            auto nbrOffset = neighbor.offset;
+            numNbrsInFirstHop++;
+            if (!visited.contains(nbrOffset)) {
+                const auto nbrVector =
+                    embeddings->getEmbedding(transaction, embeddingScanState, nbrOffset);
+                auto dist = HNSWIndexUtils::computeDistance(config.distFunc, queryVector, nbrVector,
+                    embeddings->getDimension());
+                candidatesForSecHop.push({nbrOffset, dist});
+                if (semiMask->isMasked(nbrOffset)) {
+                    if (results.size() < ef || dist < results.top().distance) {
+                        if (results.size() == ef) {
+                            results.pop();
+                        }
+                        numVisitedNbrs++;
+                        visited.add(nbrOffset);
+                        results.push({nbrOffset, dist});
+                        candidates.push({nbrOffset, dist});
+                    }
+                }
+            }
+        });
+    }
+
+    common::offset_vec_t filteredNeighbors;
+    while (!candidatesForSecHop.empty()) {
+        auto [candidate, candidateDist] = candidatesForSecHop.top();
+        candidatesForSecHop.pop();
+        if (numVisitedNbrs >= numNbrsInFirstHop) {
+            break;
+        }
+        if (visited.contains(candidate)) {
+            continue;
+        }
+        visited.add(candidate);
+        auto secondHopNbrItr =
+            lowerGraph->scanFwd(common::nodeID_t{candidate, nodeTableID}, secondHopScanState);
+        for (const auto& secondHopNbrChunk : secondHopNbrItr) {
```

Comment:
You seem to be iterating over all 2nd degree neighbors here. Instead, we should stop early when we run distance computations on M many nodes (M is the maximum degree in the lower layer). That is we continue visiting neighbors (1st degree or 2nd degree) until M many unvisited nodes that are also in the mask are seen.

Reply:
Changed it to stop early here.